Mathlib/NumberTheory/AbelSummation.lean

```diff
@@ -0,0 +1,173 @@
+private theorem integralmulsum (hf_diff : ∀ t ∈ Set.Icc a b, DifferentiableAt ℝ f t)
+    (hf_int : IntegrableOn (deriv f) (Set.Icc a b)) (t₁ t₂ : ℝ) (n : ℕ) (h : t₁ ≤ t₂)
+    (h₁ : n ≤ t₁) (h₂ : t₂ ≤ n + 1) (h₃ : a ≤ t₁) (h₄ : t₂ ≤ b) :
+    ∫ t in t₁..t₂, deriv f t * ∑ k ∈ Icc 0 ⌊t⌋₊, c k =
+      (f t₂ - f t₁) * ∑ k ∈ Icc 0 n, c k := by
+  have h_inc₁ : Ι t₁ t₂ ⊆ Set.Icc n (n + 1) :=
+    Set.uIoc_of_le h ▸ Set.Ioc_subset_Icc_self.trans <| Set.Icc_subset_Icc h₁ h₂
+  have h_inc₂ : Set.uIcc t₁ t₂ ⊆ Set.Icc a b := Set.uIcc_of_le h ▸ Set.Icc_subset_Icc h₃ h₄
+  rw [← integral_deriv_eq_sub (fun t ht ↦ hf_diff t (h_inc₂ ht)), ← integral_mul_const]
+  · refine integral_congr_ae ?_
+    filter_upwards [sumlocc c n] with t h h' using by rw [h (h_inc₁ h')]
+  · refine (intervalIntegrable_iff_integrableOn_Icc_of_le h).mpr (hf_int.mono_set ?_)
+    rwa [← Set.uIcc_of_le h]
+
+private theorem ineqofmemIco {k : ℕ} (hk : k ∈ Set.Ico (⌊a⌋₊ + 1) ⌊b⌋₊) :
+    a ≤ k ∧ k + 1 ≤ b := by
+  constructor
+  · have := (Nat.succ_eq_add_one _) ▸ (Set.mem_Ico.mp hk).1
+    exact le_of_lt <| (Nat.floor_lt' (by linarith)).mp this
+  · rw [← Nat.cast_add_one, ← Nat.le_floor_iff' (by linarith)]
+    exact (Set.mem_Ico.mp hk).2
+
+private theorem ineqofmemIco' {k : ℕ} (hk : k ∈ Ico (⌊a⌋₊ + 1) ⌊b⌋₊) :
+    a ≤ k ∧ k + 1 ≤ b := ineqofmemIco (by rwa [← Finset.coe_Ico])
+
+private theorem integrablemulsum (ha : 0 ≤ a) (hb : ⌊a⌋₊ < ⌊b⌋₊)
+  (hf_int : IntegrableOn (deriv f) (Set.Icc a b)) :
+    IntegrableOn (fun t ↦ deriv f t * (∑ k ∈ Icc 0 ⌊t⌋₊, c k)) (Set.Icc a b) := by
+  suffices h_locint : ∀ (t₁ t₂ : ℝ) (n : ℕ) (h : t₁ ≤ t₂) (h₁ : n ≤ t₁) (h₂ : t₂ ≤ n + 1)
+      (h₃ : a ≤ t₁) (h₄ : t₂ ≤ b),
+      IntervalIntegrable (fun t ↦ deriv f t * (∑ k ∈ Icc 0 ⌊t⌋₊, c k)) volume t₁ t₂ by
+    -- We prove some inequalities to be used later on by linarith
+    have : 0 < b := Nat.pos_of_floor_pos <| lt_of_le_of_lt (Nat.zero_le _) hb
+    have : ⌊a⌋₊ + 1 ≤ b := by rwa [← Nat.cast_add_one, ← Nat.le_floor_iff (by positivity)]
+    have : a < ⌊a⌋₊ + 1 := Nat.lt_floor_add_one _
+    have : a < ⌊b⌋₊ := by rwa [← Nat.floor_lt ha]
+    rw [← intervalIntegrable_iff_integrableOn_Icc_of_le (by linarith)]
+    refine (h_locint a (⌊a⌋₊ + 1 : ℕ) ⌊a⌋₊ ?_ ?_ ?_ ?_ ?_).trans <|
+      (trans_iterate_Ico ?_ fun k hk ↦ h_locint k _ k ?_ ?_ ?_ ?_ ?_).trans
+      (h_locint ⌊b⌋₊ b ⌊b⌋₊ ?_ ?_ ?_ ?_ ?_)
+    any_goals simp only [Nat.cast_add_one, le_rfl]
+    any_goals linarith [ineqofmemIco hk]
+    any_goals exact Nat.floor_le (by positivity)
+    any_goals exact (Nat.lt_floor_add_one _).le
+    any_goals linarith
+  intro _ _ n h h₁ h₂ h₃ h₄
+  rw [intervalIntegrable_iff_integrableOn_Icc_of_le h]
+  exact (IntegrableOn.mono_set (hf_int.mul_const (∑ k ∈ Icc 0 n, c k))
+    (Set.Icc_subset_Icc h₃ h₄)).congr
+      <| ae_restrict_of_ae_restrict_of_subset
+        (Set.Icc_subset_Icc h₁ h₂) <| (ae_restrict_iff' measurableSet_Icc).mpr
+          (by filter_upwards [sumlocc c n] with t h ht using by rw [h ht])
```

Comment:
I'm personally not so keen on the repeated use of `any_goals`. Some of the goals you are passing to `linarith` are utterly trivial (one of them is `b ≤ b`, for example); I think these should be dealt with as they arise (using `le_rfl`).

Here's my suggested alternative for this proof. It's actually a tiny bit longer than yours, but it compiles more than twice as fast, and I hope it's a bit more obvious how the pieces fit together in a way that might make it more maintainable in future (i.e. more robust against changes elsewhere in the library):

```suggestion
private theorem integrablemulsum (ha : 0 ≤ a) (hb : ⌊a⌋₊ < ⌊b⌋₊)
  (hf_int : IntegrableOn (deriv f) (Set.Icc a b)) :
    IntegrableOn (fun t ↦ deriv f t * (∑ k ∈ Icc 0 ⌊t⌋₊, c k)) (Set.Icc a b) := by
  have h_locint {t₁ t₂ : ℝ} {n : ℕ} (h : t₁ ≤ t₂) (h₁ : n ≤ t₁) (h₂ : t₂ ≤ n + 1)
      (h₃ : a ≤ t₁) (h₄ : t₂ ≤ b) :
      IntervalIntegrable (fun t ↦ deriv f t * (∑ k ∈ Icc 0 ⌊t⌋₊, c k)) volume t₁ t₂ := by
    rw [intervalIntegrable_iff_integrableOn_Icc_of_le h]
    exact (IntegrableOn.mono_set (hf_int.mul_const _) (Set.Icc_subset_Icc h₃ h₄)).congr
      <| ae_restrict_of_ae_restrict_of_subset (Set.Icc_subset_Icc h₁ h₂)
        <| (ae_restrict_iff' measurableSet_Icc).mpr
          (by filter_upwards [sumlocc c n] with t h ht using by rw [h ht])
  -- We prove some inequalities to be used later on
  have aux1 : 0 ≤ b := (Nat.pos_of_floor_pos <| (Nat.zero_le _).trans_lt hb).le
  have aux2 : ⌊a⌋₊ + 1 ≤ b := by rwa [← Nat.cast_add_one, ← Nat.le_floor_iff aux1]
  have aux3 : a ≤ ⌊a⌋₊ + 1 := (Nat.lt_floor_add_one _).le
  have aux4 : a ≤ ⌊b⌋₊ := le_of_lt (by rwa [← Nat.floor_lt ha])
  -- now break up into 3 subintervals
  rw [← intervalIntegrable_iff_integrableOn_Icc_of_le (aux3.trans aux2)]
  have I1 : IntervalIntegrable _ volume a ↑(⌊a⌋₊ + 1) :=
    h_locint (mod_cast aux3) (Nat.floor_le ha) (mod_cast le_rfl) le_rfl (mod_cast aux2)
  have I2 : IntervalIntegrable _ volume ↑(⌊a⌋₊ + 1) ⌊b⌋₊ :=
    trans_iterate_Ico hb fun k hk ↦ h_locint (mod_cast k.le_succ)
      le_rfl (mod_cast le_rfl) (ineqofmemIco hk).1 (mod_cast (ineqofmemIco hk).2)
  have I3 : IntervalIntegrable _ volume ⌊b⌋₊ b :=
    h_locint (Nat.floor_le aux1) le_rfl (Nat.lt_floor_add_one _).le aux4 le_rfl
  exact (I1.trans I2).trans I3
```

Reply:
Yes, I agree that your version is cleaner and today I learned about `mod_cast` which is really helpful here.